File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(vector<int> numbers) { // 1,000,000
    vector<int> answer;
    answer.assign(numbers.size(), -1); 
    
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq; // min heap {수, 인덱스}
    
    for(int i = 0; i < numbers.size(); i++){
        int n = numbers[i];
        while(!pq.empty() && pq.top().first < n){
            answer[pq.top().second] = n; 
            pq.pop(); 
        }
        pq.push({n, i}); // {수, 인덱스}
    }
    
    
    return answer; // 모든 원소에 대한 뒷 큰수들을 차례로 담은 배열
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;


class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length]; 
        Arrays.fill(answer, -1);
        
        Stack<Integer> stack = new Stack<>(); stack.add(0); 
        for(int i = 1; i < numbers.length; i++){
            int num = numbers[i]; 
            while(!stack.isEmpty()){
                if(numbers[stack.peek()] < num){
                    answer[stack.peek()] = num; 
                    stack.pop(); 
                }else{
                    break;
                }
            }
            stack.add(i);
        }
        
        
        
        // 배열 의 각 원소들에 대해 자신보다 뒤에 있는 숫자 중에서 자신보다 크면서 가장 가까이 있는 수를 뒷 큰수
        // 모든 원소에 대한 뒷 큰수들을 차례로 담은 배열을 return
        return answer;
    }
}